tests/utils/mod.rs

```diff
@@ -0,0 +1,264 @@
+                    .unwrap()
+                    .pop()
+                    .unwrap();
+                // Have the initial member create an Add proposal using the new
+                // KeyPackage.
+                let add_proposal = mls_group.create_add_proposal(
+                    group_aad,
+                    initial_credential_bundle,
+                    next_member_key_package,
+                );
+                proposal_list.push(add_proposal);
+            }
+            // Create the commit based on the previously compiled list of
+            // proposals.
+            let (commit_mls_plaintext, welcome_option, _key_package_bundle_option) = mls_group
+                .create_commit(
+                    group_aad,
+                    &initial_credential_bundle,
+                    proposal_list.clone(),
+                    false,
+                )
+                .unwrap();
+            let welcome = welcome_option.unwrap();
+            // Distribute the Welcome message to the other members.
+            for client_id in 1..group_config.members.len() {
+                let new_group_member = test_clients
+                    .get(group_config.members[client_id].name)
+                    .unwrap()
+                    .borrow_mut();
+                // Figure out which key package bundle we should use. This is
+                // a bit ugly and inefficient.
+                let member_secret = welcome
+                    .get_secrets_ref()
+                    .iter()
+                    .find(|x| {
+                        match new_group_member
+                            .key_package_bundles
+                            .borrow()
+                            .iter()
+                            .find(|y| y.get_key_package().hash() == x.key_package_hash)
+                        {
+                            Some(_) => true,
+                            None => false,
+                        }
+                    })
+                    .unwrap();
+                let kpb_position = new_group_member
+                    .key_package_bundles
+                    .borrow()
+                    .iter()
+                    .position(|y| y.get_key_package().hash() == member_secret.key_package_hash)
+                    .unwrap();
+                let key_package_bundle = new_group_member
+                    .key_package_bundles
+                    .borrow_mut()
+                    .remove(kpb_position);
+                // Create the local group state of the new member based on the
+                // Welcome.
+                let new_group =
+                    MlsGroup::new_from_welcome(welcome.clone(), None, key_package_bundle).unwrap();
+                new_group_member
+                    .group_states
+                    .borrow_mut()
+                    .insert(new_group.context().group_id.clone(), new_group);
+            }
+            // Make all members receive and process the commit message.
```

Comment:
This Commit message was issued before the members joind. Just processing the Welcome message brings them up to speed, there is no need to replay the Commit message.

Reply:
I thought the process was that new members first get the Welcome message, which would then enable them to process their own Add commit. As depicted [here](https://messaginglayersecurity.rocks/mls-protocol/draft-ietf-mls-protocol.html#section-4).

Reply:
Then either the spec is wrong or openmls is wrong, let's figure out which on it is.

Reply:
The spec is ambiguous. In [Welcoming New Members](https://messaginglayersecurity.rocks/mls-protocol/draft-ietf-mls-protocol.html#section-11.2.1), it says:
> The sender of a Commit message is responsible for sending a Welcome message to any new members added via Add proposals. The Welcome message provides the new members with the current state of the group, after the application of the Commit message. The new members will not be able to decrypt or verify the Commit message, but will have the secrets they need to participate in the epoch initiated by the Commit message.

So you're right and I don't need to send the Commit messages to the new members, but Section 4 of the Spec is wron.g.